Allow new keys & users/groups to be created in the same deploy

ryanartecona opened the pull request:
### Delay chown of keys until user/group both exist

Instead of chowning keys to their user/group every time they are sent,
only attempt the chown during send-keys if the user and group both
exist, and again do a chown during activation after the users and groups
have been created.

One result is that if a key and its user and/or group are to be created
in the same `nixops deploy`, the key will first be uploaded and owned
by root:root, then chmod'd, then late in activation the key will be
chowned to the newly created user/group. This includes a node's first
deploy, when it has neither keys nor users/groups.

Another result is that between send-keys and the next deploy (often,
but not necessarily, in the same `nixops deploy`), a key may
have its permissions set as configured, but _not_ be owned by the
configured user/group (instead root:root), which is presumed safe 
(**is it really though?**).
### Add service per key to track key dependencies

If a user service wants to block until `deployment.keys.mysecret` exists
to start up, it can now do so by adding `"mysecret-key.service"` to
its `requires` and `after` lists.

I tried doing this first via a systemd mount unit and a systemd path unit.
I couldn't get the mount to work (I have little experience configuring
filesystems). The path unit only supports unit _activation_, and only for
a single unit, and besides that it's extra fiddly to configure. 

For the per-key services, the `preStart` script will block until the key exists, 
and then the `script` will block until the key no longer exists. Permissions
aren't checked, only existence, as permissions are assumed to be correctly
set during activation.

basvandijk commented:
+1 I like it.

ip1981 commented on nix/keys.nix:
```diff
@@ -136,23 +136,64 @@ in
+          # Make sure each key has correct ownership, since the configured owning
+          # user or group may not have existed when first uploaded.
+          ''
+            [[ -f "/run/keys/${name}" ]] && chown '${value.user}:${value.group}' "/run/keys/${name}"
```
what if the file does not exist?
I'd use single quotes.

ip1981 replied:
Maybe keep only 
```
chown '${value.user}:${value.group}' '/run/keys/${name}'
```
to fail noisily

ryanartecona replied:
Perhaps. I was unsure whether to fail noisily, or to treat this as best-effort chowning of keys that both (1) we know about at this point in activation and (2) exist, having been placed here for us before activation even started.

I opted for the latter, so that the absence of one key doesn't screw up ownership of other keys or fail an activation. Also, if a service depends on a specific key, they can now do so via a systemd service dependency, since each key gets a service which is running as long as the key exists (added below).

kevincox commented:
I'll test this out today.

kevincox commented:
This is working perfectly for me! Thanks so much, this will save me a lot of manual fiddling in the near future.

ip1981 commented:
Using numbers instead of names works right now :-)

kevincox commented:
I know, but I generate my groups programmatically and assigning unique numbers would be a non-trivial task. So I'm putting up with it for now. I like not having to think about group numbers if I can avoid it.

ip1981 commented:
Ah, yeah. I did assigning unique IDs trivial: https://marc.info/?l=nix-dev&m=145046990923112&w=2

kevincox commented:
On 08/06/16 10:44, Igor Pashev wrote:
> Ah, yeah. I did assigning unique IDs trivial:
> https://marc.info/?l=nix-dev&m=145046990923112&w=2

Clever. If this doesn't ship soon I might try that out.

ryanartecona commented:
Any big reason why this can't be merged? Is it just a lack of testing?

kevincox commented:
Yeah, I was going to ping this today as well because I had to provision a
new machine.
On Jun 24, 2016 5:47 PM, "Ryan Artecona" <notifications@github.com> wrote:

> Any big reason why this can't be merged? Is it just a lack of testing?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixops/pull/400#issuecomment-228470104>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAeJvKXrecIU4tUwXB4frqJ2RcH6JZ1Eks5qPFB3gaJpZM4Hi-uE>
> .
>

ryanartecona commented:
Just rebased onto current master.

domenkozar commented:
@ryanartecona two things I think would make it easier to merge this:

- document how keys are now used in deployments to nixops manual
- provide a [simple example](sscce.org) how to test this PR with nixops

I know this is effort, but I've seen such PRs rot exactly for these reasons, reviewers don't have much time if they're not particularly interested into the feature/bug.

kevincox commented:
@ryanartecona Do you have any plans to add the documentation. I think only a couple of lines would be required.

@domenkozar For a simple example see the following. Either of the clauses will fail (when deploying to a machine where that user/group did not previously exist) where as this patch should make the deploy successful.

```nix
{
	users.users.a-user = {};
	deployment.keys.example = {
		user = "a-user";
		text = "blah";
	};
	
	# or
	
	users.groups.a-group = {};
	deployment.keys.example = {
		group = "a-group";
		text = "group blah";
	};
}
```

ryanartecona commented:
Sorry for the radio silence.

I just rebased this branch onto current master and added commit aeb5b1e which documents `deployment.keys`,  and `nixops send-keys`. I would definitely like a second pair of eyes on the documentation. It seems to build and render as intended on my computer, but I have never worked with docbook before. 

Tangentially, I noticed that the "Command Reference" section of the manual is fairly thorough, but I had never realized it because none of its subsections are in the table of contents. I made a feeble effort to get those `<refsection>`s into the table of contents, but my docbook-fu isn't nearly strong enough.

Anything else I should provide so this can get merged?

kevincox commented on doc/manual/overview.xml:
```diff
@@ -1562,6 +1562,94 @@ deploy the new configuration.
+
+  <para>
+    Since the <filename>/run</filename> directory is mounted as a tmpfs filesystem,
+    keys stored there from <varname>deployment.keys</varname> are stored in RAM
```
I prefer:

> Keys from <varname>deployment.keys</varname> are stored user <filename>/run/</filename> on a temporary filesystem and will not be persisted across a reboot. To send a rebooted machine it's keys, use <command>nixops send-keys</command>. Note that all  <command>nixops</command> commands implicitly upload keys when appropriate,  so manually sending keys should only be necessary after an unattended reboot.

It is a slightly rearranged version that
- Removes a couple implementation details that I didn't think were relevant here.
- Made `nixops send-keys` sentence a little more clear and consistent with terminology.

What do you think?

kevincox commented on doc/manual/overview.xml:
```diff
@@ -1562,6 +1562,94 @@ deploy the new configuration.
+
+  <note>
+    <para>
+      Keys in <literal>/run/keys/</literal> are not stored on disk by default.
```
I think this paragraph is unnecessary because the following one covers this.

ryanartecona replied:
I was on the fence here myself. I wasn't sure if this was important enough to give a little highlighted callout, or just to trust that whoever it concerns will read through this whole new keys section. Thinking more now, the following paragraph should be sufficient. Taking this out.

kevincox commented on doc/manual/overview.xml:
```diff
@@ -1562,6 +1562,94 @@ deploy the new configuration.
+
+    A service with a key dependency set up in this way will be prevented by systemd
+    from being started after an unattended reboot, and vanilla <command>systemctl status</command>
+    and friends will lead you to the cause.
```
How about:

> These decencies will ensure that the service is started when the keys it requires are present. For example, after a reboot the services will be delayed until the keys are available and <command>systemctl status</command>  and friends will lead you to the cause.

- avoids the ambiguous phrase "...a key dependency...
- Changes "prevents starting" to "starts when the keys are present" (sounds more positive I think)

ryanartecona commented:
@kevincox Thanks for the thoughtful docs review! I've incorporated your suggestions.

kevincox commented:
Thanks, LGTM.

@domenkozar What do you think?

spinus commented:
+1 for the work, cheers.

Would be nice to add a test for it as well.

domenkozar commented:
Sorry it took so long, but I'll try to review this shortly. Who's using this feature?

domenkozar commented:
This also fixed https://github.com/NixOS/nixops/issues/461

kevincox commented:
I'm using this feature. This fixes problems whenever someone is using a key assigned to a named group. When nixops is run, and the group does not yet exist you will get an error uploading keys because it tries to chown it to the group (which does not yet exist).

spinus commented:
I'm not using yet but I would like to :-) for the same reasons Kevin mentioned.
